### PR TITLE
use best filtering when possible and use filter when downsampling

### PR DESCRIPTION
**Pull request by SimonDanisch**

### CODE REVIEW

**Comment by SimonDanisch** [comment]
Just to get an idea of the artifacts we get with the different filter approaches:
FILTER_NEAREST with downsampling:
image

FILTER_BILINEAR:
image

FILTER_BEST:
image

File: src/primitives.jl [hunk 1]
```diff
@@ -469,6 +469,35 @@ function regularly_spaced_array_to_range(arr)
     end
 end
 
+"""
+    interpolation_flag(is_vector, scene, interp, w, h)
```
Comment (ericphanson) [inline src/primitives.jl:473]:
```suggestion
    interpolation_flag(is_vector, interp, wpx, hpx, w, h)
```

File: src/primitives.jl [hunk 2]
```diff
@@ -492,20 +521,24 @@ function draw_atomic(scene::Scene, screen::CairoScreen, primitive::Union{Heatmap
     imsize = (extrema_nan(xs), extrema_nan(ys))
 
     interp = to_value(get(primitive, :interpolate, true))
-
     weird_cairo_limit = (2^15) - 23
+
+    # Debug attriubte we can set to disable fastpath
```
Comment (ericphanson) [inline src/primitives.jl:526]:
```suggestion
    # Debug attribute we can set to disable fastpath
```

File: src/primitives.jl [hunk 3]
```diff
@@ -492,20 +521,24 @@ function draw_atomic(scene::Scene, screen::CairoScreen, primitive::Union{Heatmap
...
+    # probably shouldn't really be part of the interface
+    fast_path = to_value(get(primitive, :fastpath, true))
```
Comment (ericphanson) [inline src/primitives.jl:528]: maybe
```suggestion
    fast_path = to_value(get(primitive, :fast_path, true))
```
?

**Review by ericphanson — APPROVED** [review]

**Comment by ericphanson** [comment]
lots of method overwrite warnings in the CI logs btw

**Comment by SimonDanisch** [comment]
>lots of method overwrite warnings in the CI logs btw

Where?

**Comment by ericphanson** [comment]
Maybe only on nightly? https://github.com/JuliaPlots/CairoMakie.jl/pull/179/checks?check_run_id=2805148599#step:6:329

**Comment by SimonDanisch** [comment]
Oh wow :D Well, it's in StructArrays anyways...Guess that would need fixing there. 
I'd also much rather test on 1.5 + 1.6 instead of 1.6 + nightly

**Pull request merged**